report/sections/5-type-checking.tex

```diff
@@ -0,0 +1,230 @@
+\newpage
+\section{Type Checking}
+\label{sec:type-checking}
+
+In this section we will dicuss our attempt at using NaBL2 \citep{Antwerpen:2016:CLS:2847538.2847543} to implement the type system as explained in section \ref{sec:structural-typing-typescript}.
+NaBL2 is a constraint generation language specifically tailored towards the problem of name binding.
+The language allows language designers to specify constraints over AST constructs, which are then solved in one pass by the runtime.
+A general solution to the name binding problem is provided by a framework of scope graphs.
+A scope graph consists of scopes, declarations and references to names in a program.
+Constraints can be generated to construct such graphs, as well as to resolve paths from declarations to references in the graph.
+Since the problem of type checking is very similar to that of name binding, NaBL2 can also be used to implement a type checker.
+
+\subsection{Structural type checking}
+
+In general, NaBL2 is well-suited for statements and expressions that are common in programming languages, 
+such as the shown \textit{IfStatement}.
+During this project, our research focussed on exploring to what extend NaBL2 supports structural type checking and, 
+in the case that it does not, what would be required to support it.
+Our syntax definition for TypeScript includes several language constructs that must participate in type checking.
+These constructs included: object declarations, variable assignments, interfaces, functions, function calls and primitives.
+Several test cases and examples were developed that tested various aspects of structural type checking with these constructs.
+
+To demonstrate our attempt at implementing structural types, we will use object assignment as example.
+The code below shows a TypeScript program that should be rejected by the type checker 
+because \texttt{Person} is a structural subtype of \texttt{Developer}. 
```

Comment:
Or it is, but this sentence is rather confusing. Better to remove double negation